lib/matplotlib/figure.py

```diff
@@ -1522,6 +1523,187 @@ def subplots(self, nrows=1, ncols=1, sharex=False, sharey=False,
                 .subplots(sharex=sharex, sharey=sharey, squeeze=squeeze,
                           subplot_kw=subplot_kw))
 
+    @staticmethod
+    def _normalize_grid_string(layout):
+        layout = inspect.cleandoc(layout)
+        return [list(ln) for ln in layout.strip('\n').split('\n')]
+
+    def subplot_mosaic(self, layout, *, subplot_kw=None, gridspec_kw=None,
+                       empty_sentinel='.'):
+        """
+        Build a layout of Axes based on ASCII art or nested lists.
+
+        This is a helper function to build complex GridSpec layouts visually.
+
+        .. note ::
+
+           This API is provisional and may be revised in the future based on
+           early user feedback.
+
+
+        Parameters
+        ----------
+        layout : list of list of {hashable or nested} or str
+
+            A visual layout of how you want your Axes to be arranged
+            labeled as strings.  For example ::
+
+               x = [['A', 'A', 'edge'],
+                    ['C', '.', 'edge']]
+
+            Produces 4 axes:
+
+            - 'A' which is 1 row high and spans the first two columns
+            - 'edge' which is 2 rows high and is on the right edge
+            - 'C' which in 1 row and 1 column wide in the bottom left
+            - a blank space 1 row and 1 column wide in the bottom center
+
+            If input is a str, then it must be of the form ::
+
+              '''
+              AAE
+              C.E
+              '''
+
+            where the labels are assumed to be single characters and
+            the rows are delimited by new lines.  This only allows
+            single character axes labels and does not allow nesting,
+            but is very terse.
+
+            Using nested lists any hashable objects can be used to label
+            the Axes.  In the special case that the objects in the list are
+            existing Axes objects they are adjusted to match the requested
+            layout.
+
+        subplot_kw : dict, optional
+            Dictionary with keywords passed to the `.Figure.add_subplot` call
+            used to create each subplot.
+
+        gridspec_kw : dict, optional
+            Dictionary with keywords passed to the `.GridSpec` constructor used
+            to create the grid the subplots are placed on.
+
+        empty_sentinel : object, optional
+            Entry in the layout to mean "leave this space empty".  Defaults
+            to ``'.'``. Note, if *layout* is a string, it is processed via
+            `inspect.cleandoc` to remove leading white space, which may
+            interfere with using white-space as the empty sentinel.
+
+        Returns
+        -------
+        dict[label, Axes]
+           A dictionary mapping the labels to the Axes objects.
```

Comment:
can this be [name, Axes] - worried label will imply that that's the label we're going to apply to the axes. Unless is that the plan - these keys end up being the defaults for a latex like subfigure label?